Add DPG testserver tests for service driven evolution

fengzhou-msft opened the pull request:
# Description
This PR uses customized code to fix the dpg code of `dpg_update1` and `media_types` due to #1746
Addresses https://github.com/azure/autorest.csharp/issues/1944

Testing cases:
- [x] **params_getNewOperation** - Initially neither path or method exist for this operation. After evolution, this is a **new method in a new path**
- [x] **params_deleteParameters** - Initially the path exists but there is no delete method. After evolution this is a **new method in a known path**
- [x] **params_headNoParams** - Initially has no query parameters. After evolution, a **new optional query parameter** is added
- [x] **params_getRequired** - Initially only has one required Query Parameter. After evolution, a **new optional query parameter** is added
- [x] **params_getOptional** - Initially has one optional query parameter. After evolution, a **new optional query parameter** is added
- [x] **params_putRequiredOptional** - Initially has one required query parameter and one optional query parameter.  After evolution, a **new optional query parameter** is added

Relevant docs:
- https://github.com/Azure/autorest.testserver/wiki/Data-plane-codegen-testing#service-driven-evolution
- https://github.com/Azure/autorest.testserver/blob/main/swagger/dpg_initial.json
- https://github.com/Azure/autorest.testserver/blob/main/swagger/dpg_update1.json

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

ShivangiReja commented on test/TestServerProjectsLowLevel/llc_update1/Custom/ParamsClient.cs:
```diff
@@ -0,0 +1,81 @@
+        /// <param name="context"> The request context, which can override default behaviors on the request on a per-call basis. </param>
+        /// <exception cref="ArgumentNullException"> <paramref name="content"/> is null. </exception>
+#pragma warning disable AZC0002
+        public virtual async Task<Response> PostParametersAsync(RequestContent content, ContentType contentType, RequestContext context = null)
```
I believe we want these methods for testing purposes, I wonder why don't we add this method in testserver repo in the [swagger](https://github.com/Azure/autorest.testserver/blob/main/swagger/llc_update1.json) so that other languages can also leverage?

ShivangiReja commented on test/TestServerProjectsLowLevel/llc_update1/Custom/ParamsClient.cs:
```diff
@@ -0,0 +1,81 @@
+        /// <param name="context"> The request context, which can override default behaviors on the request on a per-call basis. </param>
+        /// <exception cref="ArgumentNullException"> <paramref name="content"/> is null. </exception>
+#pragma warning disable AZC0002
+        public virtual async Task<Response> PostParametersAsync(RequestContent content, ContentType contentType, RequestContext context = null)
```
I believe we want this method for testing purposes, I wonder why don't we add this in testserver repo in the [swagger](https://github.com/Azure/autorest.testserver/blob/main/swagger/llc_update1.json) so that other languages can also leverage this?

fengzhou-msft replied:
This method does come from testserver swagger https://github.com/Azure/autorest.testserver/blob/main/swagger/llc_update1.json#L93 and we use custom code here as LLC currently generates duplicate methods.

annelo-msft commented on test/TestServerProjectsLowLevel/llc_update1/Custom/ParamsClient.cs:
```diff
@@ -0,0 +1,81 @@
+    [CodeGenSuppress("PostParametersAsync", typeof(RequestContent), typeof(RequestContext))]
+    [CodeGenSuppress("PostParameters", typeof(RequestContent), typeof(RequestContext))]
+    [CodeGenSuppress("CreatePostParametersRequest", typeof(RequestContent), typeof(RequestContext))]
+    public partial class ParamsClient
```
@fengzhou-msft, I see you've added customization here.  Is the idea that once we've introduced "state in autorest" we'll come back and update this test to handle it fully generated?

fengzhou-msft replied:
yes, that's the plan.

annelo-msft commented on test/AutoRest.TestServerLowLevel.Tests/dpg_initial.cs:
```diff
@@ -0,0 +1,41 @@
+    public class DpgInitialTest : TestServerLowLevelTestBase
+    {
+        [Test]
+        public Task DPGAddOptionalInput() => Test(async (host) =>
```
nit: please use .NET casing convention consistently.

fengzhou-msft replied:
Updated.

annelo-msft commented on test/AutoRest.TestServerLowLevel.Tests/dpg_update1.cs:
```diff
@@ -0,0 +1,79 @@
+            var result3 = await new ParamsClient(Key, host).GetRequiredAsync("a", context: default);
+            var responseBody3 = JsonData.FromBytes(result3.Content.ToMemory());
+            Assert.AreEqual("An object was successfully returned", (string)responseBody3["message"]);
+            var result4 = await new ParamsClient(Key, host).GetRequiredAsync("a", "b");
```
Could you help me understand what's happening here?  How do we know we didn't make breaking changes between the initial test and the update1 test?

fengzhou-msft replied:
Customized code like overload methods are added to make v2(update1) backward compatible for non-breaking changes from swagger. Here all the test cases of initial are copied to update1. Therefore, we can consider update1 is backward compatible with inital.

annelo-msft commented on test/TestServerProjectsLowLevel/dpg_update1/Custom/ParamsClient.cs:
```diff
@@ -0,0 +1,184 @@
+namespace dpg_update1_LowLevel
+{
+    /// <summary> The Params service client. </summary>
+    [CodeGenSuppress("PostParametersAsync", typeof(RequestContent), typeof(RequestContext))]
```
Curious: What do the [CodeGenSuppress] attributes do?

fengzhou-msft replied:
The autogenerated `PostParametersAsync` has the issue of generating duplicate methods, so here we can use `CodeGenSuppress` to suppress the autogenerated `PostParametersAsync` methods whose input parameters are of the types of RequestContent and RequestContext. Then we can write the correct customized code for a `PostParametersAsync` method.

The generator will scan existing customized code (all the files in the project that are not under the Generated folder) for the attributes like CodeGenSuppress, then apply the actions like removing the method to the generated code with Roslyn code analyzer.

annelo-msft replied:
I see, thank you for that explanation!

Is that the final approach we decided for addressing this problem, that we'll recommend not using the generated method and writing the two method overloads for different content types by hand?

annelo-msft replied:
Per our offline conversation, we will handle this as part of a separate PR so we don't block this PR on that work.

annelo-msft reviewed (changes requested):
I've updated the description with the full set of test cases from the swagger -- I'm not sure I see tests that validate each of them, but if they're there, let me know and I'll have another look at this tomorrow.  Thanks!

fengzhou-msft commented on test/AutoRest.TestServer.Tests/Infrastructure/TestServerLowLevelTestBase.cs:
```diff
@@ -81,7 +81,7 @@ private async Task Test(string scenario, Func<Uri, Task> test, bool ignoreScenar
         {
             var testName = TestContext.CurrentContext.Test.Name;
-            var indexOfUnderscore = testName.IndexOfAny(new[]{'_', '('});
+            var indexOfUnderscore = testName.LastIndexOfAny(new[]{'_', '('});
```
Previously we have cases to use multiple tests to test different cases of the same scenario like `DpgNewBodyType_JSON` and `DpgNewBodyType_JPEG` will both count coverage for `DpgNewBodyType`. Therefore, we have this logic to drop the part after the underscore.

In the updated testserver, however, the scenario name itself can have underscores like `DpgAddOptionalInput_OptionalParam`. Therefore, I updated here to drop the part after the last underscore and we can append `_Test` (for instance, use `DpgAddOptionalInput_OptionalParam_Test`) to the test name if the test name contains underscores.

fengzhou-msft commented on test/AutoRest.TestServerLowLevel.Tests/dpg_initial.cs:
```diff
@@ -0,0 +1,41 @@
+            var result2 = await new ParamsClient(Key, host).GetRequiredAsync("a", default);
+            var responseBody2 = JsonData.FromBytes(result2.Content.ToMemory());
+            Assert.AreEqual("An object was successfully returned", (string)responseBody2["message"]);
+            var result3 = await new ParamsClient(Key, host).GetRequiredAsync("a", context: default);
```
Here I'm trying to list all the ways that a customer can call with the method in `dpg_initial` and then copy the same tests to `dpg_update1`. We can catch that some cases are not supported by `dpg_update1` generated code. After adding customized overload methods to `dpg_update1` , all the cases here in `dpg_initial` are then supported by `dpg_update1`.

annelo-msft replied:
@AlexanderSher, does this make sense to you now?

fengzhou-msft commented:
> I've updated the description with the full set of test cases from the swagger -- I'm not sure I see tests that validate each of them, but if they're there, let me know and I'll have another look at this tomorrow. Thanks!

I just noticed that there are now more scenarios supported by the testserver. I have added all the ones listed in the description.

annelo-msft commented on test/TestServerProjectsLowLevel/dpg_update1/Custom/ParamsClient.cs:
```diff
@@ -0,0 +1,320 @@
+        /// </summary>
+        /// <param name="optionalParam"> I am an optional parameter. </param>
+        /// <param name="context"> The request context, which can override default behaviors on the request on a per-call basis. </param>
+        public virtual async Task<Response> GetOptionalAsync(string optionalParam, RequestContext context)
```
Are we able to ExportAPI.cs for these clients?  It would be helpful for reviewing the public APIs for them.

annelo-msft commented on test/TestServerProjectsLowLevel/dpg_update1/Custom/ParamsClient.cs:
```diff
@@ -0,0 +1,320 @@
+        /// </summary>
+        /// <param name="optionalParam"> I am an optional parameter. </param>
+        /// <param name="context"> The request context, which can override default behaviors on the request on a per-call basis. </param>
+        public virtual async Task<Response> GetOptionalAsync(string optionalParam, RequestContext context)
```
I see, we are writing these manually right now, but in the future, these methods will be generated by the state-aware generator, is that correct?

fengzhou-msft replied:
Yes, that's my understanding.

annelo-msft commented on test/TestServerProjectsLowLevel/dpg_update1/Custom/ParamsClient.cs:
```diff
@@ -0,0 +1,320 @@
+            scope.Start();
+            try
+            {
+                using HttpMessage message = CreatePostParametersRequest(content, ContentType.ApplicationJson, context);
```
Did we decide that our convention is that if we don't take a ContentType, we default ContentType to ApplicationJson?  Why is this better than always taking a ContentType and setting the default value of ContentType to ApplicationJson?

fengzhou-msft replied:
This is based on current generated code and I agree that the backward compatible PostParametersAsync method here could cause confusion if a customer never used v1 and started with v2. We can generate the contenType parameter with default value if it only has one content type for LLC.

annelo-msft replied:
Yes, I see, this was my misunderstanding.

The update1 API looks like this:

```csharp
public virtual Response PostParameters(RequestContent content, RequestContext context = null);
public virtual Response PostParameters(RequestContent content, ContentType contentType, RequestContext context = null);
```

The initial method should make RequestContext be required, since only one overload in a method group should have optional parameters.  Let's discuss in our design meeting today whether contentType should take a default value.

fengzhou-msft replied:
Per the discussion, I added `[EditorBrowsable(EditorBrowsableState.Never)]` to the overload methods for v1 that does not have contentType parameter.

annelo-msft commented on test/AutoRest.TestServerLowLevel.Tests/dpg_initial.cs:
```diff
@@ -0,0 +1,91 @@
+            Assert.IsEmpty(result.Content.ToArray());
+            Assert.AreEqual(200, result.Status);
+            Assert.AreEqual(123, result.Headers.ContentLength);
+            var result2 = await new ParamsClient(Key, host).HeadNoParamsAsync(new Azure.RequestContext());
```
nit: would it be possible to separate test cases with carriage returns?  It would make them reasier to read.

annelo-msft commented on test/AutoRest.TestServerLowLevel.Tests/dpg_initial.cs:
```diff
@@ -0,0 +1,91 @@
+    public class DpgInitialTest : TestServerLowLevelTestBase
+    {
+        [Test]
+        public Task DpgAddOptionalInput_NoParams_Test() => Test(async (host) =>
```
What are the tests validating in dpg_initial.cs?  Is it establishing a baseline so we know the same code works on both versions?  Because in this method, we are not validating that you can add an optional parameter -- that doesn't happen until the v2 client in dpg_update1.cs.

My goal for this comment is just to make sure the test is named in a way that other developers can understand.  If we want to keep the name AddOptionalInput, let's add a comment stating where the addition happens so folks aren't confused when it's not here.

fengzhou-msft replied:
The test name has to start with DpgAddOptionalInput_NoParams which matches the name in testserver, otherwise we get a coverage report error, but we can modify the suffix after the last `_`, maybe use V1 and V2 correspondingly. I will add comments as well.

annelo-msft commented on test/AutoRest.TestServerLowLevel.Tests/dpg_initial.cs:
```diff
@@ -0,0 +1,91 @@
+        [Test]
+        public Task DpgAddOptionalInput_NoParams_Test() => Test(async (host) =>
+        {
+            var result = await new ParamsClient(Key, host).HeadNoParamsAsync();
```
Should we test this passing `default` for context as well?

fengzhou-msft replied:
Added.

annelo-msft commented on test/AutoRest.TestServerLowLevel.Tests/dpg_initial.cs:
```diff
@@ -0,0 +1,91 @@
+            var result3 = await new ParamsClient(Key, host).GetRequiredAsync("param", context: default);
+            var responseBody3 = JsonData.FromBytes(result3.Content.ToMemory());
+            Assert.AreEqual("An object was successfully returned", (string)responseBody3["message"]);
+            Assert.ThrowsAsync<ArgumentNullException>(async () => await new ParamsClient(Key, host).GetRequiredAsync(null));
```
We shouldn't be validating that the method throws on a null query parameters, only if it is a path parameter.

annelo-msft commented on test/AutoRest.TestServerLowLevel.Tests/dpg_initial.cs:
```diff
@@ -0,0 +1,91 @@
+            Assert.AreEqual("An object was successfully returned", (string)responseBody2["message"]);
+            var result3 = await new ParamsClient(Key, host).GetRequiredAsync("param", context: default);
+            var responseBody3 = JsonData.FromBytes(result3.Content.ToMemory());
+            Assert.AreEqual("An object was successfully returned", (string)responseBody3["message"]);
```
Should we add a test case for passing a non-default context?

fengzhou-msft replied:
Updated.

annelo-msft commented on test/AutoRest.TestServerLowLevel.Tests/dpg_update1.cs:
```diff
@@ -0,0 +1,171 @@
+        public Task DpgAddOptionalInput() => Test(async (host) =>
+        {
+            // compatible with dpg_initial
+            var result = await new ParamsClient(Key, host).GetRequiredAsync("param");
```
Is there a way to reuse the code from the dpg_initial to ensure that it is identical?  This feels a little error-prone if someone modified one test file and not the other.  It may not be a big deal to do it this way for tests, though.

fengzhou-msft replied:
Refactored to share dpg_initial test code.

annelo-msft reviewed (commented):
Added mostly questions, this is getting close!

annelo-msft commented:
Do we have any test cases covering these scenarios? https://gist.github.com/annelo-msft/827658b363941fdde699ac3925dd6ccf#2-changing-an-input-that-was-required-in-vi-to-optional-in-vi1

If not, we'll want to create an issue to do this in a separate PR.

fengzhou-msft commented on test/AutoRest.TestServer.Tests/Infrastructure/TestServerLowLevelTestBase.cs:
```diff
@@ -81,8 +81,13 @@ private async Task Test(string scenario, Func<Uri, Task> test, bool ignoreScenar
+            var indexOfDelimiter = testName.LastIndexOf('_');
+            if (indexOfDelimiter == -1)
+            {
+                indexOfDelimiter = testName.IndexOf('(');
```
Previously, test names do not contain `_`.  Now the testserver has `DPGAddOptionalInput_NoParams`, `DPGAddOptionalInput_RequiredOptionalParam`, `DPGAddOptionalInput_OptionalParam` as test names. Instead of passing `DPGAddOptionalInput` as the name to report coverage, now we need to pass the full name. 

As a workaround to not break previous tests, I made this change and we will name the test like `DpgAddOptionalInput_NoParams_Test` and then `DpgAddOptionalInput_NoParams` will be passed as the name for coverage to match the full test name.

fengzhou-msft commented on test/AutoRest.Shared.Tests/AutoRest.Shared.Tests.csproj:
```diff
@@ -20,4 +20,4 @@
     <Compile Include="../../src/assets/**/*.cs" />
```
This looks like to be caused by editor after opening the file, the last empty line is removed. I'll revert the change.
